### ISSUE TITLE
[F#] Implement new Concept Exercise: memory-allocation

### ISSUE THREAD

**Issue by ErikSchierboom** [exercism/v3#1153]
This issue describes how to implement the `memory-allocation` concept exercise for the F# track.

## Goal

The goal of this exercise is to teach the student the basics of the Concept of Memory Allocation in F#.

## Learning objectives

- Know what the stack and heap are.
- Know what the garbage collector is.
- Know how the stack and heap are different.
- Know what the large object heap is.
- Know what reference and value types are.
- Know how reference and value types relate to the stack and heap.
- Know how to make a discriminated union, record, tuple or class a struct.

## Out of scope

- Memory locality.
- Memory pooling.
- Pointer semantics.
- `Span<T>` and `Memory<T>`
- Readonly semantics and by reference arguments.
- Discussion of profiling tools (referring to them is fine).

## Concepts

The Concepts this exercise unlocks are:

- `memory-allocation`: know what the stack and heap are; know what the garbage collector is; know how the stack and heap are different; know what the large object heap is; know what reference and value types are; know how reference and value types relate to the stack and heap; know how to make a discriminated union, record, tuple or class a struct.

## Prerequisites

This exercise's prerequisites Concepts are:

- `classes`: know how to define a class.
- `records`: know how to define a record.
- `tuples`: know how to define a tuple.
- `discriminated-unions`: know how to define a discriminated union.
- `attributes`: know how to add attributes.

Any data types used in this exercise (e.g. `strings` or `integers`) should also be added as prerequisites.

### Hints

- [Structs][structs]: describes what structs are and how to define a type (class) as a struct
- [Struct tuples][struct-tuples]: how to define struct tuples.
- [Struct records][struct-records]: how to define struct records.
- [Struct discriminated unions][struct-discriminated-unions]: how to define struct discriminated unions.

The following links all use C# as their language, but the concepts apply equally as well to F#:

- [Value types][value-types]: introduction to value types.
- [Reference types][reference-types]: introduction to reference types.
- [Stack versus heap][stack-vs-heap]: how the stack and heap work.

### After

- [Structs][structs]: describes what structs are and how to define a type (class) as a struct
- [Struct tuples][struct-tuples]: how to define struct tuples.
- [Struct records][struct-records]: how to define struct records.
- [Struct discriminated unions][struct-discriminated-unions]: how to define struct discriminated unions.

The following links all use C# as their language, but the concepts apply equally as well to F#:

- [Value types][value-types]: introduction to value types.
- [Reference types][reference-types]: introduction to reference types.
- [Stack versus heap][stack-vs-heap]: how the stack and heap work.
- [Value types vs reference types][value-types-vs-reference-types]: in-depth comparison between value and reference types
- [Garbage collection][garbage-collection]: how garbage collection works in C#.
- [Large Object Heap][large-object-heap]: what the Large Object Heap is.
- [Garbage collection performance][garbage-collection-performance]: in-depth look into garbage collection performance.

## Representer

This exercise does not require any specific representation logic to be added to the [representer][representer].

## Analyzer

This exercise does not require any specific logic to be added to the [analyzer][analyzer].

## Implementing

To implement the concept exercise, the following files must be created:

<pre>
languages
└── fsharp
    └── exercises
        └── concept
            └── memory-allocation
                ├── .docs
                |   ├── instructions.md
                |   ├── introduction.md
                |   ├── hints.md
                |   └── after.md
                ├── .meta
                |   ├── config.json
                |   |── design.md
                |   └── Example.fs
                ├── MemoryAllocation.fs
                ├── MemoryAllocation.fsproj
                └── MemoryAllocationTests.fs
</pre>

## Step 1: add .docs/introduction.md

This file contains an introduction to the concept. It should make the exercise's learning goals explicit and provide a short introduction with enough detail for the student to complete the exercise. The aim is to give the student just enough context to figure out the solution themselves, as research has shown that self-discovery is the most effective learning experience. Using the proper technical terms in the descriptions will be helpful if the student wants to search for more information. If the exercise introduces new syntax, an example of the syntax should always be included; students should not need to search the web for examples of syntax.

As an example, the introduction to a "strings" exercise might describe a string as just a "Sequence of Unicode characters" or a "series of bytes". Unless the student needs to understand more nuanced details in order to solve the exercise, this type of brief explanation (along with an example of its syntax) should be sufficient information for the student to solve the exercise.

## Step 2: add .docs/instructions.md

This file contains instructions for the exercise. It should explicitly explain what the student needs to do (define a method with the signature `X(...)` that takes an A and returns a Z), and provide at least one example usage of that function. If there are multiple tasks within the exercise, it should provide an example of each.

## Step 3: add .docs/hints.md

If the student gets stuck, we will allow them to click a button requesting a hint, which shows this file. This will not be a "recommended" path and we will (softly) discourage them using it unless they can't progress without it. As such, it's worth considering that the student reading it will be a little confused/overwhelmed and maybe frustrated.

The file should contain both general and task-specific "hints". These hints should be enough to unblock almost any student. They might link to the docs of the functions that need to be used.

The hints should not spell out the solution, but instead point to a resource describing the solution (e.g. linking to documentation for the function to use).

## Step 4: add .docs/after.md

Once the student completes the exercise they will be shown this file, which should provide them with a summary of what the exercise aimed to teach. This document can also link to any additional resources that might be interesting to the student in the context of the exercise.

These files are also all described in the [concept exercises document][docs-concept-exercises].

## Step 5: update languages/fsharp/config.json

An entry should be added to the track's `config.json` file for the new concept exercise:

```json
{
  ...
  "exercises": {
    "concept": [
      ...
      {
        "slug": "memory-allocation",
        "uuid": "912068eb-2d47-43f8-add3-5c4ef3d2b8b1",
        "concepts": ["memory-allocation"],
        "prerequisites": [
          "classes",
          "records",
          "tuples",
          "discriminated-unions",
          "attributes"
        ]
      }
    ]
  }
}
```

## Step 6: adding track-specific files

These files are specific to the F# track:

- `MemoryAllocation.fs`. the stub implementation file, which is the starting point for students to work on the exercise.
- `MemoryAllocation.fsproj`: the F# project file.
- `MemoryAllocationTests.fs`: the test suite.
- `.meta/Example.fs`: an example implementation that passes all the tests. The project file should _not_ include this file.

## Step 7: update the general concept document

Add the exercise to the concept's shared document's `## Implementations` section ([example][reference-example]). As there currently is no such reference document, a stub version of that file should be created at `reference/concepts/memory-allocation.md`.

## Step 8: updating list of implemented exercises

- Add the exercise to the [list of implemented exercises][implemented-exercises].

## Step 9: add .meta/design.md:

This file contains information on the exercise's design, which includes things like its goal, its teaching goals, what not to teach, and more ([example][meta-design]). This information can be extracted from this GitHub issue.

## Step 10: add .meta/config.json:

This file contains meta information on the exercise, which currently only includes the exercise's contributors ([example][meta-config.json]).

## Help

If you have any questions while implementing the exercise, please post the questions as comments in this issue.

[how-to-implement-a-concept-exercise]: https://github.com/exercism/v3/blob/master/docs/maintainers/generic-how-to-implement-a-concept-exercise.md
[implemented-exercises]: https://github.com/exercism/v3/tree/master/languages/fsharp/exercises/concept/README.md#implemented-exercises
[reference]: https://github.com/exercism/v3/blob/master/reference/concepts/memory-allocation.md
[reference-example]: https://github.com/exercism/v3/blob/master/reference/types/string.md#implementations
[analyzer]: https://github.com/exercism/fsharp-analyzer
[representer]: https://github.com/exercism/fsharp-representer
[meta-design]: https://github.com/exercism/v3/blob/master/languages/fsharp/exercises/concept/numbers/.meta/design.md
[meta-config.json]: https://github.com/exercism/v3/blob/master/languages/fsharp/exercises/concept/numbers/.meta/config.json
[docs-concept-exercises]: https://github.com/exercism/v3/blob/master/docs/concept-exercises.md
[value-types]: https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/builtin-types/value-types
[reference-types]: https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/reference-types
[stack-vs-heap]: https://www.c-sharpcorner.com/article/C-Sharp-heaping-vs-stacking-in-net-part-i/
[garbage-collection]: https://docs.microsoft.com/en-us/dotnet/standard/garbage-collection/fundamentals
[value-types-vs-reference-types]: https://adamsitnik.com/Value-Types-vs-Reference-Types/
[large-object-heap]: https://docs.microsoft.com/en-us/dotnet/standard/garbage-collection/large-object-heap
[garbage-collection-performance]: https://docs.microsoft.com/en-us/dotnet/standard/garbage-collection/performance
[struct-tuples]: https://docs.microsoft.com/en-us/dotnet/fsharp/language-reference/tuples#syntax
[struct-records]: https://docs.microsoft.com/en-us/dotnet/fsharp/language-reference/records#remarks
[struct-discriminated-unions]: https://docs.microsoft.com/en-us/dotnet/fsharp/language-reference/discriminated-unions#struct-discriminated-unions
[structs]: https://docs.microsoft.com/en-us/dotnet/fsharp/language-reference/structures
